### PR TITLE
Migrations: Implement inheritance

### LINKED ISSUE #1706: Migrations: Implement inheritance support
- Update the differ to understand TPH.
- ~~Consider gen'ing a check constraint for discriminator value set.~~ (Created #2594)

### PR DESCRIPTION
**Pull request by bricelam**
Resolves #1706

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs [hunk 1]
```diff
@@ -310,5 +310,12 @@ public static IKey GetKey([NotNull] this IEntityType entityType, [NotNull] IRead
 
             return key;
         }
+
+        public static IEnumerable<IKey> GetDeclaredKeys([NotNull] this IEntityType entityType)
+        {
+            Check.NotNull(entityType, nameof(entityType));
+
+            return entityType.GetKeys().Where(k => k.EntityType == entityType);
+        }
```
Comment (bricelam) [inline src/EntityFramework.Core/Metadata/EntityTypeExtensions.cs]: @AndriySvyryd Any reason this one was "missed"? Does anything else need to be done in this area?

File: src/EntityFramework.Core/Metadata/EntityType.cs [hunk 2]
```diff
@@ -305,7 +305,6 @@ public virtual Key AddKey([NotNull] Property property)
         public virtual Key AddKey([NotNull] IReadOnlyList<Property> properties)
         {
             Check.NotEmpty(properties, nameof(properties));
-            ThrowIfDerivedEntity();
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: We need to be careful about expanding the scope of inheritance for RTM. If supporting keys on non-root types is something we think is high priority we need to make sure it works across the stack (query, fixup, code first, etc.)

Comment (bricelam) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: @anpete @ajcvickers @divega @rowanmiller Thoughts? Should we scope to not allow alternate keys on derived types? Code First appeared to superficially work after removing this.

Comment (rowanmiller) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: I'm ok with limiting it for now and enabling it when we have time to test it properly

Comment (divega) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: +1

Comment (bricelam) [inline src/EntityFramework.Core/Metadata/EntityType.cs]: Will undo. Filed #2611.

**Comment by AndriySvyryd** [comment]
:shipit:

**Pull request merged**